engine/src/main/java/org/terasology/rendering/dag/WorldReflectionNode.java

```diff
@@ -0,0 +1,105 @@
+import org.terasology.rendering.world.RenderQueuesHelper;
+import org.terasology.rendering.world.WorldRenderer;
+import org.terasology.rendering.world.WorldRendererImpl;
+
+import static org.lwjgl.opengl.GL11.GL_COLOR_BUFFER_BIT;
+import static org.lwjgl.opengl.GL11.GL_DEPTH_BUFFER_BIT;
+import static org.lwjgl.opengl.GL11.glClear;
+import static org.terasology.rendering.opengl.OpenGLUtils.bindDisplay;
+
+/**
+ * Diagram of this node can be viewed from:
+ * TODO: move diagram to the wiki when this part of the code is stable
+ * - https://docs.google.com/drawings/d/1Iz7MA8Y5q7yjxxcgZW-0antv5kgx6NYkvoInielbwGU/edit?usp=sharing
+ */
+public class WorldReflectionNode implements Node {
+
+    private RenderQueuesHelper renderQueues;
+    private RenderingConfig renderingConfig;
+    private WorldRenderer worldRenderer;
+    private Material chunkShader;
+
+    private Context context;
+    private BackdropRenderer backdropRenderer;
+    private Camera playerCamera;
+
+    private FBO sceneReflected;
+    private FBO sceneOpaque;
+
+    public WorldReflectionNode(Context context, Camera playerCamera, Material chunkShader) {
+        this.context = context;
+        this.backdropRenderer = context.get(BackdropRenderer.class);
+        this.playerCamera = playerCamera;
+        this.renderingConfig = context.get(Config.class).getRendering();
+        this.chunkShader = chunkShader;
+        this.worldRenderer = context.get(WorldRenderer.class);
+        this.renderQueues = context.get(RenderQueuesHelper.class);
+    }
+
+    @Override
+    public void update(float deltaInSeconds) {
+        this.sceneReflected = context.get(FrameBuffersManager.class).getFBO("sceneReflected");
+        this.sceneOpaque = context.get(FrameBuffersManager.class).getFBO("sceneOpaque");
+    }
+
+    @Override
+    public void process() {
+        PerformanceMonitor.startActivity("Render World (Reflection)");
+
+        sceneReflected.bind();
+        OpenGLUtils.setViewportToSizeOf(sceneReflected);
+        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+        GL11.glCullFace(GL11.GL_FRONT);
+        playerCamera.setReflected(true);
+
+
+        playerCamera.lookThroughNormalized(); // we don't want the reflected scene to be bobbing or moving with the player
+        backdropRenderer.render(playerCamera);
+        playerCamera.lookThrough();
+
+        if (renderingConfig.isReflectiveWater()) {
+            chunkShader.activateFeature(ShaderProgramFeature.FEATURE_USE_FORWARD_LIGHTING);
+            worldRenderer.renderChunks(renderQueues.chunksOpaqueReflection, ChunkMesh.RenderPhase.OPAQUE, playerCamera, WorldRendererImpl.ChunkRenderMode.REFLECTION);
+            chunkShader.deactivateFeature(ShaderProgramFeature.FEATURE_USE_FORWARD_LIGHTING);
+        }
+
+        playerCamera.setReflected(false);
+
+        GL11.glCullFace(GL11.GL_BACK);
+        bindDisplay();
+        OpenGLUtils.setViewportToSizeOf(sceneOpaque); // toWholeDisplay
```

Comment:
Let's remove "OpenGLUtils." 

Regarding the //toWholeDisplay comment, indeed the original code had "setViewportToWholeDisplay();"  Why did you change it? 

Interestingly, this is one case in which a state change is used also later, outside of the render reflection step. In the following rendering steps, i.e. backdrop render and then the whole scene opaque, the viewport stays to what is set here. Again, state management will have to take care of this: each node should be independent of the others unless there is an explicit requirement (I'm thinking a node could have a setExecutionDependency(Node ancestor) method - or something like that), state management will then ensure that the GL state is changed only if necessary.